File: tobas_gcs/src/tobas_gcs/apps/simulation/wind_parameters.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ...gcs import GroundControlStationWidget

import math
import rospy
from PyQt5.QtCore import Qt, pyqtSlot
from PyQt5.QtWidgets import QWidget, QLabel, QVBoxLayout
from PyQt5.QtGui import QFont

from tobas_rqt_tools.widgets import FloatSliderTextWidget
from tobas_rqt_tools.layouts import FormLayout
from tobas_rqt_tools.messages import q_error

from tobas_tools_py.drone import Drone
from tobas_gazebo_msgs.srv import (
    GetWindParams,
    GetWindParamsRequest,
    GetWindParamsResponse,
    SetWindParams,
    SetWindParamsRequest,
    SetWindParamsResponse,
)

from ...common import TITLE_PSIZE


class WindParamsWidget(QWidget):
    WAIT_FOR_SERVICE = 30.0  # [s]

    def __init__(self, main: GroundControlStationWidget, drone: Drone) -> None:
        super().__init__()

        self._main = main
        self._drone = drone

        self._get_sc = None
        self._set_sc = None

        rows = QVBoxLayout()
        self.setLayout(rows)

        title = QLabel("Wind Parameters")
        title.setFont(QFont("Default", pointSize=TITLE_PSIZE, weight=QFont.Bold))
        title.setAlignment(Qt.AlignTop)
        rows.addWidget(title)

        form = FormLayout()
        rows.addLayout(form)

        self._mean_speed = FloatSliderTextWidget(0.0, 20.0)
        self._mean_speed.value_changed.connect(self._on_value_changed)
        form.addRow(QLabel("Mean Speed [m/s]"), self._mean_speed)

        self._direction = FloatSliderTextWidget(-math.pi, math.pi)
        self._direction.value_changed.connect(self._on_value_changed)
        form.addRow(QLabel("Direction [rad]"), self._direction)

        self._gust_speed_factor = FloatSliderTextWidget(0.0, 10.0)
        self._gust_speed_factor.value_changed.connect(self._on_value_changed)
        form.addRow(QLabel("Gust Speed Factor [-]"), self._gust_speed_factor)

        self._gust_duration = FloatSliderTextWidget(0.0, 10.0)
        self._gust_duration.value_changed.connect(self._on_value_changed)
        form.addRow(QLabel("Gust Duration [s]"), self._gust_duration)

        self._gust_interval = FloatSliderTextWidget(0.0, 30.0)
        self._gust_interval.value_changed.connect(self._on_value_changed)
        form.addRow(QLabel("Gust Interval [s]"), self._gust_interval)

    def initialize(self) -> bool:
        self._get_sc = rospy.ServiceProxy(f"{self._drone.name}/gazebo/get_wind_parameters", GetWindParams)
        self._set_sc = rospy.ServiceProxy(f"{self._drone.name}/gazebo/set_wind_parameters", SetWindParams)

        try:
            self._get_sc.wait_for_service(rospy.Duration(self.WAIT_FOR_SERVICE))
            self._set_sc.wait_for_service(rospy.Duration(self.WAIT_FOR_SERVICE))
        except rospy.ROSException:
            q_error(self._main, "Failed to connect to Gazebo wind server.")
            return False

        # パラメータの初期値を設定
        self._load_current_params()

        return True

    @pyqtSlot()
    def _on_value_changed(self) -> None:
        set_params_req = SetWindParamsRequest()
        set_params_req.params.mean_speed = self._mean_speed.get()
        set_params_req.params.direction = self._direction.get()
        set_params_req.params.gust_speed_factor = self._gust_speed_factor.get()
        set_params_req.params.gust_duration = self._gust_duration.get()
        set_params_req.params.gust_interval = self._gust_interval.get()

        set_params_res: SetWindParamsResponse = self._set_sc.call(set_params_req)
        if not set_params_res.success:
            q_error(self._main, "Failed to set wind parameters.")
            self._load_current_params()
            return

    def _load_current_params(self) -> None:
        get_params_res: GetWindParamsResponse = self._get_sc.call(GetWindParamsRequest())
        cur_params = get_params_res.params

        self._mean_speed.set(cur_params.mean_speed)
        self._direction.set(cur_params.direction)
        self._gust_speed_factor.set(cur_params.gust_speed_factor)
        self._gust_duration.set(cur_params.gust_duration)
        self._gust_interval.set(cur_params.gust_interval)

File: tobas_gui/tobas_simulation_gui/src/dynamic_configuration/wind_parameters.cpp
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Tobas, Inc.

#include "tobas_simulation_gui/dynamic_configuration/wind_parameters.hpp"

#include <QDebug>
#include <QHBoxLayout>
#include <QVBoxLayout>

#include <tobas_gazebo_common/constants.hpp>
#include <tobas_gui_common/constants.hpp>
#include <tobas_path_tools/join.hpp>
#include <tobas_qt_tools/layouts/form_layout.hpp>
#include <tobas_qt_tools/message.hpp>
#include <tobas_qt_tools/widgets/label.hpp>
#include <tobas_std_tools/unit_conversions.hpp>

#include "tobas_simulation_gui/dynamic_configuration/constants.hpp"

namespace ch = std::chrono;

namespace tobas
{
namespace gui
{
namespace sim
{
WindParamsWidget::WindParamsWidget()
{
  const auto title = new qt::Label("Wind Parameters", cmn::kLabelPSize, QFont::Bold);

  reset_button_ = new QPushButton("Reset");
  reset_button_->setFixedSize(kHeaderButtonWidth, kHeaderButtonHeight);

  mean_speed_ = new qt::DoubleSliderTextWidget(0.0, 20.0, 1);
  direction_ = new qt::IntSliderTextWidget(-180, 180);
  gust_speed_factor_ = new qt::DoubleSliderTextWidget(0.0, 10.0, 1);
  gust_duration_ = new qt::DoubleSliderTextWidget(0.0, 10.0, 1);
  gust_interval_ = new qt::DoubleSliderTextWidget(0.0, 30.0, 1);

  // Layout
  const auto header_cols = new QHBoxLayout();
  header_cols->addWidget(title);
  header_cols->addStretch();
  header_cols->addWidget(reset_button_);

  const auto form = new qt::FormLayout();
  form->addVAlignedRow("Mean Speed [m/s]", mean_speed_);
  form->addVAlignedRow("Direction [deg]", direction_);
  form->addVAlignedRow("Gust Speed Factor [-]", gust_speed_factor_);
  form->addVAlignedRow("Gust Duration [s]", gust_duration_);
  form->addVAlignedRow("Gust Interval [s]", gust_interval_);

  const auto rows = new QVBoxLayout();
  rows->addLayout(header_cols);
  rows->addLayout(form);

  setLayout(rows);

  // Connection
  connect(reset_button_, &QPushButton::clicked, this, &self::onResetButtonClicked);
  connect(mean_speed_, &qt::DoubleSliderTextWidget::valueChanged, this, &self::onValueChanged);
  connect(direction_, &qt::IntSliderTextWidget::valueChanged, this, &self::onValueChanged);
  connect(gust_speed_factor_, &qt::DoubleSliderTextWidget::valueChanged, this, &self::onValueChanged);
  connect(gust_duration_, &qt::DoubleSliderTextWidget::valueChanged, this, &self::onValueChanged);
  connect(gust_interval_, &qt::DoubleSliderTextWidget::valueChanged, this, &self::onValueChanged);

  reset();
}

void WindParamsWidget::reset()
{
  setParamsToDefault();
}

void WindParamsWidget::initializeRosInterfaces(rclcpp::Node::SharedPtr node, const std::string& ns)
{
  get_sc_ = std::make_shared<ros2::SyncServiceClient<GetSrv>>(node, path::join(ns, gazebo::kGetWindParamsSrv));
  set_sc_ = std::make_shared<ros2::SyncServiceClient<SetSrv>>(node, path::join(ns, gazebo::kSetWindParamsSrv));
}

void WindParamsWidget::clearRosInterfaces()
{
  get_sc_.reset();
  set_sc_.reset();
}

double WindParamsWidget::getMeanSpeed() const
{
  return mean_speed_->get();
}

double WindParamsWidget::getDirection() const
{
  return st::deg2rad(direction_->get());
}

double WindParamsWidget::getGustSpeedFactor() const
{
  return gust_speed_factor_->get();
}

double WindParamsWidget::getGustDuration() const
{
  return gust_duration_->get();
}

double WindParamsWidget::getGustInterval() const
{
  return gust_interval_->get();
}

void WindParamsWidget::setMeanSpeed(double value)
{
  QSignalBlocker speed(mean_speed_);
  mean_speed_->set(value);
}

void WindParamsWidget::setDirection(double value_rad)
{
  QSignalBlocker speed(direction_);
  direction_->set(st::rad2deg(value_rad));
}

void WindParamsWidget::setGustSpeedFactor(double value)
{
  QSignalBlocker speed(gust_speed_factor_);
  gust_speed_factor_->set(value);
}

void WindParamsWidget::setGustDuration(double value)
{
  QSignalBlocker speed(gust_duration_);
  gust_duration_->set(value);
}

void WindParamsWidget::setGustInterval(double value)
{
  QSignalBlocker speed(gust_interval_);
  gust_interval_->set(value);
}

void WindParamsWidget::setParamsToDefault()
{
  setMeanSpeed(0.0);
  setDirection(0.0);
  setGustSpeedFactor(1.0);
  setGustDuration(5.0);
  setGustInterval(10.0);
}

bool WindParamsWidget::sendGuiParams()
{
  const auto req = std::make_shared<SetSrv::Request>();
  req->params.mean_speed = getMeanSpeed();
  req->params.direction = getDirection();
  req->params.gust_speed_factor = getGustSpeedFactor();
  req->params.gust_duration = getGustDuration();
  req->params.gust_interval = getGustInterval();

  const auto res = set_sc_->sendRequestAndWait(req);
  if (!res) {
    qt::qErrorBox(this, "Failed to call \"" + QString(gazebo::kSetWindParamsSrv) + "\" service.");
    return false;
  }

  if (!res->success) {
    qt::qErrorBox(this, "Failed to set wind parameters.");
    return false;
  }

  return true;
}

void WindParamsWidget::onResetButtonClicked()
{
  setParamsToDefault();
  sendGuiParams();
}

void WindParamsWidget::onValueChanged()
{
  sendGuiParams();
}
}  // namespace sim
}  // namespace gui
}  // namespace tobas
